File: algorithmic_toolbox/week_4/03_divide_and_conquer_starter_files_20160804/inversions/inversions.cpp
#include <iostream>
#include <vector>

using std::vector;

long long get_inversions(vector<int> &a, vector<int> &b, size_t left, int ave, size_t right) {
    size_t i = left, j = ave, k = left;
    long long inv_count = 0;
    while (i <= ave - 1 && j <= right) {
        if (a[i] <= a[j]) {
            b[k] = a[i];
            i++;
        } else {
            b[k] = a[j];
            inv_count += ave - i;
            j++;
        }
        k++;
    }
    while (i <= ave - 1) {
        b[k] = a[i];
        i++;
        k++;
    }
    while (j <= right) {
        b[k] = a[j];
        j++;
        k++;
    }
    for (i = left; i <= right; i++) {
        a[i] = b[i];
    }
    return inv_count;
}

long long merge_sort(vector<int> &a, vector<int> &b, size_t left, size_t right) {
    long long inv_count = 0;
    if (right <= left) return inv_count;
    size_t ave = left + (right - left) / 2;
    inv_count += merge_sort(a, b, left, ave);
    inv_count += merge_sort(a, b, ave + 1, right);
    inv_count += get_inversions(a, b, left, ave + 1, right);
    return inv_count;
}

int main() {
    int n;
    std::cin >> n;
    vector<int> a(n);
    for (size_t i = 0; i < a.size(); i++) {
        std::cin >> a[i];
    }
    vector<int> b(a.size());
    std::cout << merge_sort(a, b, 0, a.size() - 1) << '\n';
}

File: algorithmic_toolbox/week_4/03_divide_and_conquer_starter_files_20160804/inversions/inversions.py
# Uses python3
import sys


def merge(a, b, left, ave, right):
    inv_count = 0
    i, j, k = left, ave, left
    while i <= ave - 1 and j <= right:
        if a[i] <= a[j]:
            b[k] = a[i]
            i += 1
        else:
            b[k] = a[j]
            j += 1
            inv_count += ave - i
        k += 1
    while i <= ave - 1:
        b[k] = a[i]
        i += 1
        k += 1
    while j <= right:
        b[k] = a[j]
        j += 1
        k += 1
    for i in range(left, right + 1):
        a[i] = b[i]
    return inv_count


def merge_sort(a, b, left, right):
    inv_count = 0
    if right > left:
        ave = (left + right) // 2
        inv_count += merge_sort(a, b, left, ave)
        inv_count += merge_sort(a, b, ave + 1, right)

        inv_count += merge(a, b, left, ave + 1, right)

    return inv_count


if __name__ == '__main__':
    input = sys.stdin.read()
    n, *a = list(map(int, input.split()))
    b = n * [0]
    print(merge_sort(a, b, 0, len(a) - 1))
